src/include/migraphx/op/roialign.hpp

```diff
@@ -0,0 +1,321 @@
+            const float xx =
+                roi_start_w + pw * bin_size_w + (ix + .5f) * bin_size_w / roi_bin_grid_w;
+
+            float x = (coord_trans_mode == "output_half_pixel") ? (xx - 0.5f) : xx;
+            float y = (coord_trans_mode == "output_half_pixel") ? (yy - 0.5f) : yy;
+
+            // deal with: inverse elements are out of feature map boundary
+            if(y < -1.0 || y > height || x < -1.0 || x > width)
+            {
+                auto& pc = pos_weights[index];
+                pc.pos1  = 0;
+                pc.pos2  = 0;
+                pc.pos3  = 0;
+                pc.pos4  = 0;
+                pc.w1    = 0;
+                pc.w2    = 0;
+                pc.w3    = 0;
+                pc.w4    = 0;
+                return;
+            }
+
+            y          = (y <= 0) ? 0 : y;
+            x          = (x <= 0) ? 0 : x;
+            auto y_low = static_cast<int64_t>(y);
+            auto x_low = static_cast<int64_t>(x);
+            int64_t y_high;
+            int64_t x_high;
+
+            y_high = y_low + 1;
+            if(y_low >= height - 1)
+            {
+                y = y_high = y_low = height - 1;
+            }
+
+            x_high = x_low + 1;
+            if(x_low >= width - 1)
+            {
+                x = x_high = x_low = width - 1;
+            }
+
+            float ly = y - y_low;
+            float lx = x - x_low;
+            float hy = 1.0f - ly;
+            float hx = 1.0f - lx;
+            float w1 = hy * hx;
+            float w2 = hy * lx;
+            float w3 = ly * hx;
+            float w4 = ly * lx;
+
+            // save weights and indeces
+            pos_weight pc;
+            pc.pos1            = y_low * width + x_low;
+            pc.pos2            = y_low * width + x_high;
+            pc.pos3            = y_high * width + x_low;
+            pc.pos4            = y_high * width + x_high;
+            pc.w1              = w1;
+            pc.w2              = w2;
+            pc.w3              = w3;
+            pc.w4              = w4;
+            pos_weights[index] = pc;
+        });
+    }
+
+    struct max_pool
+    {
+        double init() { return std::numeric_limits<double>::lowest(); }
+
+        double op(double x, double y)
+        {
+            double m = std::max(x, y);
+            return (m);
+        }
+
+        double final(double x, std::size_t) { return (x); }
+    };
+
+    struct avg_pool
+    {
+        double init() { return 0.0; }
+
+        double op(double x, double y) { return x + y; }
+
+        double final(double x, std::size_t y) { return (y == 0) ? 0.0 : (x / y); }
+    };
+
+    template <class T, class Op>
+    double calc_pooling(const T* data,
+                        int64_t roi_bin_grid_h,
+                        int64_t roi_bin_grid_w,
+                        const std::vector<pos_weight>& pos_weights,
+                        int64_t& index,
```

Comment:
Return the index and output as a tuple instead of using an input parameter. You can use `std::tie` to assign each value to the vector.